src/components/_proxied-dot-io/common/docs-page/lib/get-canonical-url.ts

```diff
@@ -0,0 +1,47 @@
+import { HOSTNAME_MAP } from 'constants/hostname-map'
+
+function getDomainFromProductSlug(productSlug: string) {
+  // the end result here is a single object entry: [['www.waypointproject.io', 'waypoint']]
+  const [productDomain] = Object.entries(HOSTNAME_MAP)
+    // get all matching entries for the product
+    .filter(([, product]) => product === productSlug)
+    // ignore test domains
+    .filter(([hostname]) => !hostname.includes('hashi-mktg.com'))
+
+  // grab the domain, which is the first item of the matching entry: ['www.waypointproject.io', 'waypoint']
+  return productDomain?.[0]
+}
+
+/**
+ * Generate a canonical URL for an arbitrary product docs page
+ *
+ * Example:
+ * ```ts
+ * getCanonicalUrlForDocsPage({ baseRoute: 'docs', currentPath: 'upgrade/compatibility', productSlug: 'waypoint' })
+ * // returns https://www.waypointproject.io/docs/upgrade/compatibility
+ */
+export function getCanonicalUrlForDocsPage({
+  baseRoute,
+  currentPath,
+  productSlug,
+}: {
+  baseRoute: string
+  currentPath: string
+  productSlug: string
+}) {
+  const canonicalUrl = new URL(
```

Comment:
⛏️ Nitpick: could we construct `const canonicalUrl` after determining `productDomain` (and possibly returning early)... in order to avoid the "stub in domain"?

Reply:
Yeah, good call!